### PR TITLE
Aob 97 dispatch an event on job execution creation

### PR DESCRIPTION
**Pull request by AymericPlanche**
**Description (for Contributor and Core Developer)**

Fire an event when job execution is created

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Bundle/BatchBundle/EventListener/LoggerSubscriber.php [hunk 1]
```diff
@@ -47,6 +47,7 @@ public function __construct(LoggerInterface $logger, TranslatorInterface $transl
     public static function getSubscribedEvents()
     {
         return array(
+            EventInterface::JOB_EXECUTION_CREATED      => 'jobExecutionCreated',
```
Comment (phaseinducer) [inline src/Akeneo/Bundle/BatchBundle/EventListener/LoggerSubscriber.php]: If this class is modified its spec must be too.

File: src/Akeneo/Bundle/BatchBundle/Launcher/SimpleJobLauncher.php [hunk 2]
```diff
@@ -60,7 +67,8 @@ public function __construct(
         JobParametersValidator $jobParametersValidator,
         $rootDir,
         $environment,
-        $logDir
+        $logDir,
+        EventDispatcherInterface $eventDispatcher
```
Comment (phaseinducer) [inline src/Akeneo/Bundle/BatchBundle/Launcher/SimpleJobLauncher.php]: Please put class-typed arguments before primitive-typed arguments.

File: src/Akeneo/Bundle/BatchQueueBundle/Command/PublishJobToQueueCommand.php [hunk 3]
```diff
@@ -133,6 +138,18 @@ protected function execute(InputInterface $input, OutputInterface $output)
         return self::EXIT_SUCCESS_CODE;
     }
 
+    /**
+     * Trigger event linked to JobExecution
+     *
+     * @param string       $eventName    Name of the event
+     * @param JobExecution $jobExecution Object to store job execution
+     */
+    private function dispatchJobExecutionEvent($eventName, JobExecution $jobExecution)
```
Comment (phaseinducer) [inline src/Akeneo/Bundle/BatchQueueBundle/Command/PublishJobToQueueCommand.php]: Adding a `: void` is ok here imo : )

File: src/Akeneo/Bundle/BatchQueueBundle/Launcher/QueueJobLauncher.php [hunk 4]
```diff
@@ -43,28 +46,35 @@ class QueueJobLauncher implements JobLauncherInterface
     /** @var string */
     private $environment;
 
+    /** @var EventDispatcherInterface */
+    protected $eventDispatcher;
```
Comment (phaseinducer) [inline src/Akeneo/Bundle/BatchQueueBundle/Launcher/QueueJobLauncher.php]: `private` if you want to be consistent.
Extra blank line below also.

File: src/Akeneo/Bundle/BatchQueueBundle/Launcher/QueueJobLauncher.php [hunk 5]
```diff
@@ -43,28 +46,35 @@ class QueueJobLauncher implements JobLauncherInterface
...
      */
     public function __construct(
         JobRepositoryInterface $jobRepository,
         JobParametersFactory $jobParametersFactory,
         JobRegistry $jobRegistry,
         JobParametersValidator $jobParametersValidator,
         JobExecutionQueueInterface $queue,
-        string $environment
+        string $environment,
+        EventDispatcherInterface $eventDispatcher
```
Comment (phaseinducer) [inline src/Akeneo/Bundle/BatchQueueBundle/Launcher/QueueJobLauncher.php]: Same remark about params order.

File: src/Akeneo/Bundle/BatchBundle/EventListener/LoggerSubscriber.php (hunk 1 above)
Comment (AymericPlanche) [reply src/Akeneo/Bundle/BatchBundle/EventListener/LoggerSubscriber.php]: fixed here https://github.com/akeneo/pim-community-dev/pull/7738/commits/f8ed3d025fa5f5165b0054c80c49c4f11a752325

File: src/Akeneo/Bundle/BatchBundle/Launcher/SimpleJobLauncher.php [hunk 6]
```diff
@@ -42,22 +45,27 @@ class SimpleJobLauncher implements JobLauncherInterface
...
-     * @param string                 $environment
-     * @param string                 $logDir
+     * @param JobRepositoryInterface    $jobRepository
+     * @param JobParametersFactory      $jobParametersFactory
+     * @param JobRegistry               $jobRegistry
+     * @param JobParametersValidator    $jobParametersValidator
+     * @param string                    $rootDir
+     * @param string                    $environment
+     * @param string                    $logDir
+     * @param EventDispatcherInterface  $eventDispatcher
```
Comment (phaseinducer) [inline src/Akeneo/Bundle/BatchBundle/Launcher/SimpleJobLauncher.php]: The phpdoc and property declaration should follow the params order.

**Comment by skeleton** [comment]
if you can squash some of your commits (the 3 last one for instance), it's better ;)

File: src/Akeneo/Bundle/BatchBundle/spec/Launcher/SimpleJobLauncherSpec.php [hunk 7]
```diff
@@ -22,16 +26,49 @@ function let(
...
-        $this->beConstructedWith($jobRepository, $jobParametersFactory, $jobRegistry, $jobParametersValidator, '/', 'prod', '/logs');
+        $this->beConstructedWith($jobRepository, $jobParametersFactory, $jobRegistry, $jobParametersValidator, $eventDispatcher, '/', 'prod', '/logs');
     }
 
     function it_is_a_job_launcher()
     {
         $this->shouldHaveType(JobLauncherInterface::class);
     }
 
+    function it_launches_a_job_by_pushing_it_to_the_queue(
```
Comment (ahocquard) [inline src/Akeneo/Bundle/BatchBundle/spec/Launcher/SimpleJobLauncherSpec.php]: you are not pushing it to the queue. You are executin with a `exec`. We should not have any notion of the queue in this bundle.

File: src/Akeneo/Bundle/BatchQueueBundle/Command/PublishJobToQueueCommand.php [hunk 8]
```diff
@@ -122,6 +125,8 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         $this->getJobExecutionQueue()->publish($jobExecutionMessage);
 
+        $this->dispatchJobExecutionEvent(EventInterface::JOB_EXECUTION_CREATED, $jobExecution);
```
Comment (ahocquard) [inline src/Akeneo/Bundle/BatchQueueBundle/Command/PublishJobToQueueCommand.php]: Shouldn't it put 2 lines above? Pushing into the queue is done after the creation of a job execution and your event is about creation of a job execution.

**Pull request merged**